### PR TITLE
DEVCVL-251 - [FE] <CreateAssetPage/> Page Component

### PR DESCRIPTION
**Pull request by pracheebapate**
ss-asset-add-edit-2022-05-04_14-12

### CODE REVIEW

File: src/pages/assets/single/CreateEdit.js [hunk 1]
```diff
@@ -0,0 +1,525 @@
...
+
+    }, [collection]);
+
+    useEffect( () => {
+
+        async function process() {
+            if(nft && nft.length > 0) {
+                cvlNft = {
+                    'id':nft[0].id,
+                    'name':nft[0].attributes.name,
```
Comment (shakil-adv) [inline src/pages/assets/single/CreateEdit.js:414]: Please use `get()` method to retrieves attributes

File: src/pages/assets/single/CreateEdit.js [hunk 2]
```diff
@@ -0,0 +1,525 @@
...
+
+        process();
+
+    }, [collection]);
+
+    useEffect( () => {
+
+        async function process() {
+            if(nft && nft.length > 0) {
+                cvlNft = {
```
Comment (shakil-adv) [inline src/pages/assets/single/CreateEdit.js:412]: Please refer to https://apteam.atlassian.net/browse/DEVCVL-250 regarding the shape of our `CvlNfts` collection. `name`, `description` and `imageUrl` should be in a IPFS json file and the link of the file should be in `tokenURI` key field.

**Review by kyawthetkt — COMMENTED** [review]
@pracheebapate 
After creating an asset,you will have to update the nftIds column of CvlCollection table.
And one more thing, please add assetType(image or video or mp3) in CvlNfts table to easily determine mp3 or image or mp4 on Card Component.

File: src/pages/assets/single/CreateEdit.js (hunk 2 above)
Comment (pracheebapate) [reply src/pages/assets/single/CreateEdit.js:412]: @shakil-adv As discussed today, we will not store metadata json file on CvlNfts table. Keeping name, description, imageUrl columns on CvlNfts table.

File: src/pages/assets/single/CreateEdit.js (hunk 1 above)
Comment (pracheebapate) [reply src/pages/assets/single/CreateEdit.js:432]: Done.

File: src/pages/assets/single/CreateEdit.js [hunk 3]
```diff
@@ -0,0 +1,525 @@
...
+    // );
+
+    useEffect( () => {
+
+        async function process() {
+
+            if(collection && collection.length > 0) {
+                cvlCollection = {
+                    'id':collection[0].id,
+                    'name':collection[0].attributes.name,
```
[the remark this answers is not in the document]
Comment (pracheebapate) [reply src/pages/assets/single/CreateEdit.js]: Done.

File: src/App.js [hunk 4]
```diff
@@ -91,6 +92,14 @@ const App = () => (
                                 path="/collection/create"
                                 element={<CreateCollectionPage />}
                             />
+                            <Route
+                                path="/collection/:collectionName/asset/create"
+                                element={<CreateEditAsset />}
+                            />
+                            <Route
+                                path="/collection/:collectionName/asset/:assetName/edit"
```
[the remark this answers is not in the document]
Comment (pracheebapate) [reply src/App.js:100]: @shakil-adv Currently 
Add new asset => collection/:collectioName/asset/create
Edit exisitng asset => collection/:collectioName/asset/:assetName/edit

Examples:
Assume we have a collection named "My Colors" and it has an asset named "light green"

Add new asset => collection/my-colors/asset/create
Edit existing asset => collection/my-colors/asset/light-green/edit

I have kept paths similar to opensea.

**Comment by kyawthetkt** [comment]
@pracheebapate If it has file ext, it is not necessary to add assetType column.Thanks

**Comment by pracheebapate** [comment]
> @pracheebapate If it has file ext, it is not necessary to add assetType column.Thanks

Great then no need to add a new column for asset type. Thanks

File: src/pages/collections/single/Create.js [hunk 5]
```diff
@@ -596,7 +596,7 @@ const CreateCollectionPage = () => {
         // console.log('moralis bannerFile', moralisBannerFile);
 
         // Prepare data to insert into the database
-        const name = values.name.trim();
```
Comment (jibrail-idris) [inline src/pages/collections/single/Create.js:599]: Why force the `name` to be lowercase?

File: src/utils/utils.js [hunk 6]
```diff
@@ -116,30 +116,63 @@ export const chains = [
...
-        return "Image must be smaller than 10MB!";
+    if (!isSizeOk) {
+        return `Image must be smaller than ${size}MB!`;
     }
 
     return true;
 };
 
+export const isValidVideo = (file, extensionTypes, extensionNames, size = 100) => {
+
```
Comment (jibrail-idris) [inline src/utils/utils.js:149]: What's this video check for? If the video is uploaded to ipfs, then it shouldn't matter.

Comment (pracheebapate) [reply src/utils/utils.js:149]: This validation check 2 things:
1. file extensions .mp4, .webm, .mp3, .wav, 
2. file size 100mb

File: src/pages/collections/single/Create.js (hunk 5 above)
Comment (pracheebapate) [reply src/pages/collections/single/Create.js:599]: In today's call with Jibrail Idris we decided to follow points:

1 ) remove name validations

removed unique validation for name for asset
removed unique validation for name for collection

2) for collection URLs
collection/create
collection/my-colors
collection/my-colors-v1
collection/my-colors-v2

3) for asset URLs
asset/assetid
Or
assets/contractaddress/tokenid

Comment (pracheebapate) [reply src/pages/collections/single/Create.js:599]: @jibrail-idris Done. Changes are implemented as per our call.

**Review by jibrail-idris — APPROVED** [review]
@pracheebapate Thanks for writing the `generateCollectionName` function algorithm, It's clean and easy to read.

**Comment by kyawthetkt** [comment]
@pracheebapate 
After creating asset, CvlCollection nftIds column is updated with wrong CvlNfts pointer. 
So I am not able to view CvlNftCollections on moralis admin side.
image

File: src/databases/moralis/useCvlNftData.js [hunk 7]
```diff
@@ -10,10 +10,15 @@ const useCvlNftData = () => {
     const auctionObject = useNewMoralisObject("CvlAuctions");
 
     const create = async (params) => {
-        save(params, {
+
+        let cvlNftId, status, errorMessage;
+
+        await save(params, {
             onSuccess: (cvlNft) => {
+                cvlNftId = cvlNft.id;
```
Comment (kyawthetkt) [inline src/databases/moralis/useCvlNftData.js:18]: `cvlNftId = cvlNft.id` should be `cvlNftId = cvlNft` to be a pointer.

Comment (pracheebapate) [reply src/databases/moralis/useCvlNftData.js:18]: @shakil-adv @kyawthetkt 

Column CvlCollections.nftIds, what we actually need to store in this column?

nft ids , ['FbGueDSa1xOXfb2UP7zvY5kN', 'n33D0nXp6u8KE9Sl3AZIOqtP'] 

or

nft record pointers  
[
  {
    "__type": "Pointer",
    "className": "CvlNfts",
    "objectId": "FbGueDSa1xOXfb2UP7zvY5kN"
  },
  {
    "__type": "Pointer",
    "className": "CvlNfts",
    "objectId": "n33D0nXp6u8KE9Sl3AZIOqtP"
  }
]

Lets discuss this. Thanks

Comment (pracheebapate) [reply src/databases/moralis/useCvlNftData.js]: @shakil-adv @kyawthetkt I have changed it to an array of pointers.

File: src/pages/collections/single/Create.js [hunk 8]
```diff
@@ -539,10 +542,67 @@ const CreateCollectionPage = () => {
...
+    }
+
     const onFinish = async(values) => {
 
+        // Show processing modal
+        setLoading(true);
+
         console.log('Received values of form: ', values);
 
+        // Check collection name
```
[the remark this answers is not in the document]
Comment (pracheebapate) [reply src/pages/collections/single/Create.js:603]: Done. Corrected.

**Comment by pracheebapate** [comment]
> @pracheebapate After creating asset, CvlCollection nftIds column is updated with incorrrect CvlNfts pointer. So I am not able to view CvlNftCollections on moralis admin side and not displaying recently created assets of a collection on our side. image

@kyawthetkt This is resolved.

**Review by kyawthetkt — APPROVED** [review]

**Review by shakil-adv — APPROVED** [review]

**Pull request merged**